src/Microsoft.Health.Checkpoints/Storage/BaseStorageCheckpointClient.cs

```diff
@@ -0,0 +1,90 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using EnsureThat;
+
+namespace Microsoft.Health.Checkpoints.Storage
+{
+    public class BaseStorageCheckpointClient : ICheckpointClient
+    {
+        private BlobContainerClient _storageClient;
+
+        public BaseStorageCheckpointClient(BlobContainerClient containerClient)
+        {
+            EnsureArg.IsNotNull(containerClient);
+
+            _storageClient = containerClient;
+        }
+
+        public virtual Task<ICheckpoint> GetCheckpointAsync(string checkpointIdentifier)
+        {
+            Task<ICheckpoint> GetCheckpointAsync()
+            {
+                ICheckpoint checkpoint = new Checkpoint();
+
+                foreach (BlobItem blob in _storageClient.GetBlobs(traits: BlobTraits.Metadata, states: BlobStates.All, prefix: checkpointIdentifier, cancellationToken: CancellationToken.None))
```

Comment:
This is a foreach because there isn't just a GetBlob API call? If so the expectation is we should only have one blob found correct? In that case we should either have an error condition if more than one is found or at least a warning that is logged.

Also, is there a reason we aren't using the GetBlobsAsync?  We should use the async call when possible so the thread pool thread isn't blocked during I/O and returned to the pool for other work.

Reply:
There does not happen to be a GetBlob API call. Updated with logging and changed to GetBlobsAsync